### PR TITLE
Cleanup and improvements to board subgrids and addition of "won" grid checks

### PR DESCRIPTION
**Pull request by bast0006**

### CODE REVIEW

File: library/board.py [hunk 1]
```diff
@@ -19,40 +19,79 @@ def __init__(self) -> None:
         self.game_over = False
         self.turn_one = True
 
+    def check_board_victory(self) -> typing.Optional[str]:
+        """Determine if the entire board has been won, and by who."""
+        grid_states = np.array(
+            (self.check_subboard_victory(i) for i in "789456123"), dtype=np.str_
+        )
+        # This might have off ordering, TODO on testing it
```
Comment (A5rocks) [inline library/board.py:27]: Make sure to test this :p

File: library/board.py [hunk 2]
```diff
@@ -19,40 +19,79 @@ def __init__(self) -> None:
...
+        return self.check_grid_victory(subgrid)
+
+    @staticmethod
+    def check_grid_victory(grid: npt.NDArray[np.str_]) -> typing.Optional[str]:
+        """Determine if a given 3x3 grid has been won, and by who."""
+        for player in ("X", "O"):
+            # this turns our grid of ["X", "O", "•"] into [True, False, False] by applying the condition to every spot
+            ticks = grid == player
+
+            # Summing over True treats True as 1, False as 0. 3 Trues = 3, so "won" rows are 3.
```
Comment (A5rocks) [inline library/board.py]: *summing over booleans

File: library/board.py (hunk 1 above)
Comment (bast0006) [reply library/board.py:27]: That happens later, when we're able to set the actual items in the grids :P

File: library/board.py [hunk 3]
```diff
@@ -19,40 +19,79 @@ def __init__(self) -> None:
...
+            "3": (third, third),
+            "4": (second, first),
+            "5": (second, second),
+            "6": (second, third),
+            "7": (first, first),
+            "8": (first, second),
+            "9": (first, third),
+        }
+
+        # Mypy cannot penetrate the slicing behavior here, so we assert that it is correct
```
Comment (A5rocks) [inline library/board.py]: Not sure "penetrate" is the right word here... Something like "figure out" might work?

Actually, looking into this:

`npt.NDArray` is an alias for `np.ndarray[Any, np.dtype[ScalarType]]`

This means: `npt.NDArray[np.str_]` == `np.ndarray[Any, np.dtype[np.str_]]`... And `__getitem__` for that seems to be a TODO: https://github.com/numpy/numpy/blob/89c80ba606f4346f8df2a31cfcc0e967045a68ed/numpy/__init__.pyi#L1202-L1203

Maybe rewrite the comment to contain that information?

**Review by A5rocks — APPROVED** [review]

**Pull request merged**